py/modsys: Add sys.implementation.mpy to give supported .mpy version.

dpgeorge opened the pull request:
When importing a .mpy file it can fail if the version/flags/arch is not correct.  When you get such an error it's hard to tell exactly what went wrong and what .mpy version was expected.

One way to improve this is to give more detailed error messages, like "expecting version 5 but got version 4".  That would cost quite a bit in code size to make it useful enough.

The approach in this PR is different: to provide a `sys.implementation.mpy` entry which returns a bytes object that is the header of the .mpy file that is supported by the system being run.  This allows to programatically determine the supported .mpy file, and also for the user to find it out by inspecting this value.

Eg:
```python
MicroPython v1.11-525-g162016ad9-dirty on 2019-10-30; linux version
Use Ctrl-D to exit, Ctrl-E for paste mode
>>> import sys
>>> sys.implementation
(name='micropython', version=(1, 11, 0), mpy=b'M\x05\n\x00 ')
>>> sys.implementation.mpy
b'M\x05\n\x00 '
>>> sys.implementation.mpy[0] # initial 'M'
77
>>> sys.implementation.mpy[1] # version number
5
>>> sys.implementation.mpy[2] # flags
11
>>> bin(_)
'0b1011' # this is an x64 arch with cached-bc-lookup and unicode strings
```

Not sure if the 4th and 5th bytes are necessary (small int size, qstr window size respectively).

TODO: write test, work out code-size increase

dpgeorge committed:
py/modsys: Report .mpy version in sys.implementation.

This commit adds a sys.implementation.mpy entry when the system supports
importing .mpy files.  This entry is a 16-bit integer which encodes two
bytes of information from the header of .mpy files that are supported by
the system being run: the second and third bytes, .mpy version, and flags
and native architecture.  This allows determining the supported .mpy file
dynamically by code, and also for the user to find it out by inspecting
this value.  It's further possible to dynamically detect if the system
supports importing .mpy files by `hasattr(sys.implementation, 'mpy')`.

dpgeorge commented:
@stinos any idea why AppVeyor failed here?

stinos commented:
Looks like a preprocessor bug/conformance problem: it doesn't like the `#if MICROPY_PERSISTENT_CODE_LOAD` within `MP_DEFINE_ATTRTUPLE` which is a macro itself. I tried some things, but only workaround is not using MP_DEFINE_ATTRTUPLE and spelling out the definition of the mp_rom_obj_tuple_t manually, or move the `#if ...` condition oustide of the definition and define 2 versions.

Reproduction:
```
#define STUFF(n, ...) \
  DoIt( n, __VA_ARGS__ );

#define BAR (0)

void Foo()
{
      STUFF(0,
#if BAR
           1,
#endif
           2)
}
```

```
error C2121: '#': invalid character: possibly the result of a macro expansion
error C2059: syntax error: 'if'
```

I submitted a bug report but AFAIK the compiler team already knows, I read an article recently stating they were fixing all kins of `__VA_ARGS__` releated bugs.

dpgeorge commented:
> it doesn't like the `#if MICROPY_PERSISTENT_CODE_LOAD` within `MP_DEFINE_ATTRTUPLE` which is a macro itself.

Ahh, I see, I totally missed that it was a macro within a macro.  I'll try to fix it.

dpgeorge commented:
Ok, I fixed the macro issue.  And added a simple test.

Code size change is:
```
   bare-arm:    +0 +0.000% 
minimal x86:    +0 +0.000% 
   unix x64:  +192 +0.038% [incl +32(data)]
unix nanbox:  +128 +0.029% [incl +32(data)]
      stm32:   +40 +0.011% PYBV10
     cc3200:   +40 +0.022% 
    esp8266:   +40 +0.006% GENERIC
      esp32:   +48 +0.004% GENERIC[incl +48(data)]
        nrf:    +0 +0.000% pca10040
       samd:    +0 +0.000% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```
Not sure why it's so big a change on unix...

stinos commented:
That size change is a bit strange yes, could be worth trying to figure out by adding code piece by piece.

For the rest: wold be goof to have this. Maybe shorten commit message header to 'Report .mpy version in sys.implementation.' if you like short messages :)

dpgeorge committed:
py/persistentcode: Move declarations for .mpy header from .c to .h file.

dpgeorge commented:
I changed the `sys.implementation.mpy` entry from a bytes object to a simple 16-bit integer, which encodes just the two most important bytes of information (second and third, version and flags and arch).  IMO this is still usable and leads to a much smaller code-size change:
```
   bare-arm:    +0 +0.000% 
minimal x86:    +0 +0.000% 
   unix x64:   +56 +0.011% 
unix nanbox:   +40 +0.009% 
      stm32:   +20 +0.005% PYBV10
     cc3200:   +16 +0.009% 
    esp8266:   +20 +0.003% GENERIC
      esp32:   +16 +0.001% GENERIC[incl +16(data)]
        nrf:    +0 +0.000% pca10040
       samd:    +0 +0.000% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```
Also note that with this new field it's possible to dynamically detect if the system supports importing .mpy files by `hasattr(sys.implementation, 'mpy')`

dpgeorge commented:
> Maybe shorten commit message header to 'Report .mpy version in sys.implementation.' if you like short messages :)

Done!

stinos commented:
Wrt nested definitions: turns out this is actually non-standard both in C (e.g. https://gcc.gnu.org/onlinedocs/cpp/Directives-Within-Macro-Arguments.html) and C++ (https://timsong-cpp.github.io/cppwp/n4659/cpp.replace#11) but gcc allows it.

gmos commented:
Found mpy in version 1.15 implementation, but not yet in the 1.15 documentation.
